hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestSnapshotChain.java

```diff
@@ -0,0 +1,305 @@
+ * limitations under the License.
+ */
+package org.apache.hadoop.ozone.om;
+
+import org.apache.hadoop.hdds.conf.OzoneConfiguration;
+import org.apache.hadoop.hdds.utils.db.Table;
+import org.apache.hadoop.ozone.om.helpers.SnapshotInfo;
+import org.apache.hadoop.util.Time;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.UUID;
+
+import static org.apache.hadoop.ozone.om.OMConfigKeys.OZONE_OM_DB_DIRS;
+
+/**
+ * Tests SnapshotChain that stores in chronological order
+ * Ozone object storage snapshots.  There exist two types
+ * of chains provided by the SnapshotChainManager
+ * i.) path based snapshots - a list of snapshots taken for a given bucket
+ * ii.) global snapshots - a list of every snapshot taken in chrono order
+ */
+public class TestSnapshotChain {
+  private OMMetadataManager omMetadataManager;
+  private Map<String, SnapshotInfo> sinfos;
+  private SnapshotChainManager chainManager;
+
+  @Rule
+  public TemporaryFolder folder = new TemporaryFolder();
+
+  @Before
+  public void setup() throws Exception {
+    OzoneConfiguration conf = new OzoneConfiguration();
+    conf.set(OZONE_OM_DB_DIRS,
+        folder.getRoot().getAbsolutePath());
+    omMetadataManager = new OmMetadataManagerImpl(conf);
+    sinfos = new HashMap<>();
+    chainManager = new SnapshotChainManager(omMetadataManager);
+  }
+
+  private SnapshotInfo createSnapshotInfo(String volName,
+                                          String bucketName,
+                                          String snapshotName,
+                                          String snapshotID,
+                                          String pathPrevID,
+                                          String globalPrevID) {
+    return new SnapshotInfo.Builder()
+        .setSnapshotID(snapshotID)
+        .setName(snapshotName)
+        .setVolumeName(volName)
+        .setBucketName(bucketName)
+        .setSnapshotStatus(SnapshotInfo.SnapshotStatus.SNAPSHOT_ACTIVE)
+        .setCreationTime(Time.now())
+        .setDeletionTime(-1L)
+        .setPathPreviousSnapshotID(pathPrevID)
+        .setGlobalPreviousSnapshotID(globalPrevID)
+        .setSnapshotPath(String.join("/", volName, bucketName))
+        .setCheckpointDir("checkpoint.testdir")
+        .build();
+  }
+
+  private void deleteSnapshot(String snapshotID) throws IOException {
+    SnapshotInfo sinfo = null;
+    final String snapshotPath = "vol1/bucket1";
+    // reset the next snapshotInfo.globalPreviousSnapshotID
+    // to the previous in the entry to be deleted.
+    if (chainManager.hasNextGlobalSnapshot(snapshotID)) {
+      sinfo = sinfos
+              .get(chainManager.nextGlobalSnapshot(snapshotID));
```

Comment:
I guess updating the  next pointers for SnapshotInfo  should also be done in case of addSnapshots  similar to what is done here for delete . Also I guess we could have separate methods define for doing this (updating pointers for SnapshotInfo ) to provide some abstraction.

Reply:
I ran the test by replacing `deleteSnapshot(snapshotID2)` by  `chainManager.deleteSnapshot(sinfos.get(snapshotID2))` and the `testDeleteSnapshot() `seems to pass here so I'm not sure why updating pointers is necessary for delete too.

Reply:
Thanks for checking the `deleteSnapshot `in the` TestSnapshotChain` @sadanand48 .  I've updated the `deleteSnapshot` private test method so it updates the` sinfos` for deleted snapshots by both updating the adjacent snapshot info pointers to point to the right snapshots in the chain _and also_ to append the deleted snapshot to the test `sinfos` data-structure for the tests.  With this, tests that would test inserting 2 of 3 snapshots, deleting a snapshot and adding a snapshot should work without issues as well as the implemented `testDeleteSnapshot()` method without having to manually update the test `sinfos` structure within the test.